gettext-rs/src/macros.rs

```diff
@@ -1,50 +1,87 @@
+/// This is an implementation detail for replacing arguments in the gettext macros. Don't call this directly.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! freplace {
+    ($input:expr, $($args:tt)*) => {{
+        let mut parts = $input.split("{}");
+        let mut output = parts.next().unwrap_or_default().to_string();
+        $(
+            output += &format!("{}{}", $args, &parts.next().expect("Argument count has to match number of format directivies ({})"));
+        )*
+        output
+    }};
+}
 
 /// Like [gettext](fn.gettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! gettext {
-    ($($arg:tt)*) => ($crate::gettext(std::fmt::format(format_args!($($arg)*))))
+    ($format:expr, $($args:expr)*) => {{
+        let s = $crate::gettext($format);
+        freplace!(s, $($args)*)
+    }};
 }
 
-/// Like [dgettext](fn.dgettext.html), but allows for formatting.
+/// Like [dgettext](fn.gettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! dgettext {
-    ($domain:expr, $string:expr, $($string_names:ident = $string_values:expr),*) => ($crate::dgettext($domain, format!($string, $( $string_names = $string_values ),*)));
+    ($domain:expr, $format:expr, $($args:expr)*) => {{
+        let s = $crate::dgettext($domain, $format);
+        freplace!(s, $($args)*)
+    }};
 }
 
 /// Like [dcgettext](fn.gettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! dcgettext {
-    ($domain:expr, $category:expr, $string:expr, $($string_names:ident = $string_values:expr),*,) => ($crate::dcgettext($domain, format!($string, $( $string_names = $string_values ),*), $category));
+    ($domain:expr, $category:expr, $format:expr, $($args:expr)*) => {{
+        let s = $crate::dcgettext($domain, $format, $category);
```

Comment:
This swaps the argument order - category is passed before the format and args string since it's a bit odd first having a vararg style input of arguments only to then pass the domain as the last argument, does that work for you?

Reply:
Yeah, I think that's better than any other alternative. :+1: